8325252: C2 SuperWord: refactor the packset

eme64 opened the pull request:
I'm refactoring the packset, separating the details of packset-manupulation from the SuperWord algorithm.

Most importantly: I split it into two classes: `PairSet` and `PackSet`.
`combine_pairs_to_longer_packs` converts the first into the second.

I was able to simplify the combining, and remove the pack-sorting.
I now walk "pair-chains" directly with `PairSetIterator`. One such pair-chain is equivalent to a pack.

I moved all the `filter / split` functionality to the `PackSet`, which allows hiding a lot of packset-manipulation from the SuperWord algorithm.

I ran into some issues when I was extending the pairset in `extend_pairset_with_more_pairs_by_following_use_and_def`:
Using the PairSetIterator changed the order of extension, and that messed with the packing heuristic, and quite a few examples did not vectorize, because we would pack up the wrong 2 nodes out of a choice of 4 (e.g. we would pack `ac bd` instead of `ab cd`). Hence, I now still have to keep the insertion order for the pairs, and this basically means we are extending with a BFS order. Maybe this issue can be removed, if I improve the packing heuristic with some look-ahead expansion approach (but that is for another day [JDK-8309908](https://bugs.openjdk.org/browse/JDK-8309908)).

But since I already spent some time on some of the packing heuristic (reordering and cost estimate), I did a light refactoring, and added extra tests for MulAddS2I.

More details are described in the annotations in the code.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325252](https://bugs.openjdk.org/browse/JDK-8325252): C2 SuperWord: refactor the packset (**Sub-task** - P4)


### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18276/head:pull/18276` \
`$ git checkout pull/18276`

Update a local copy of the PR: \
`$ git checkout pull/18276` \
`$ git pull https://git.openjdk.org/jdk.git pull/18276/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18276`

View PR using the GUI difftool: \
`$ git pr show -t 18276`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18276.diff">https://git.openjdk.org/jdk/pull/18276.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18276#issuecomment-2009802122)

eme64 commented on src/hotspot/share/opto/superword.hpp:
```diff
@@ -427,48 +546,36 @@ class SuperWord : public ResourceObj {
   // Ensure that the main loop vectors are aligned by adjusting the pre loop limit.
   void adjust_pre_loop_limit_to_align_main_loop_vectors();
-  // Is the use of d1 in u1 at the same operand position as d2 in u2?
-  bool opnd_positions_match(Node* d1, Node* u1, Node* d2, Node* u2);
```
Note: renamed, and moved up: `order_inputs_of_uses_to_match_def_pair`

eme64 commented on src/hotspot/share/opto/superword.hpp:
```diff
@@ -427,48 +546,36 @@ class SuperWord : public ResourceObj {
   // Find the longest type in def-use chain for packed nodes, and then compute the max vector size.
   int max_vector_size_in_def_use_chain(Node* n);
-  // Are s1 and s2 in a pack pair and ordered as s1,s2?
-  bool in_packset(Node* s1, Node* s2);
-  // Remove the pack at position pos in the packset
-  void remove_pack_at(int pos);
```
Note: removed

eme64 commented on src/hotspot/share/opto/superword.hpp:
```diff
@@ -427,48 +546,36 @@ class SuperWord : public ResourceObj {
   // Initialize per node info
   void initialize_node_info();
-  // Compute max depth for expressions from beginning of block
-  void compute_max_depth();
```
Note: removed.

eme64 commented on src/hotspot/share/opto/superword.hpp:
```diff
@@ -279,143 +505,36 @@ class SuperWord : public ResourceObj {
-  // Find the set of alignment solutions for load/store pack.
   const AlignmentSolution* pack_alignment_solution(const Node_List* pack);
-  // Compress packset, such that it has no nullptr entries.
-  void compress_packset();
-  // Construct the map from nodes to packs.
-  void construct_my_pack_map();
```
Note: removed

eme64 commented on src/hotspot/share/opto/superword.hpp:
```diff
@@ -279,143 +505,36 @@ class SuperWord : public ResourceObj {
   void split_packs_only_implemented_with_smaller_size();
   void split_packs_to_break_mutual_dependence();
 
-  // Filter out packs with various filter predicates
-  template <typename FilterPredicate>
-  void filter_packs(const char* filter_name,
-                    const char* error_message,
-                    FilterPredicate filter);
```
Note: moved to `PackSet`

eme64 commented on src/hotspot/share/opto/superword.hpp:
```diff
@@ -279,143 +505,36 @@ class SuperWord : public ResourceObj {
-    Node_List* second_pack() const { return _second_pack; }
-  };
-
-  SplitStatus split_pack(const char* split_name, Node_List* pack, SplitTask task);
-  template <typename SplitStrategy>
-  void split_packs(const char* split_name, SplitStrategy strategy);
```
Note: moved to `PackSet`

eme64 commented on src/hotspot/share/opto/superword.hpp:
```diff
@@ -279,143 +505,36 @@ class SuperWord : public ResourceObj {
   // do s1 and s2 have similar input edges?
   bool have_similar_inputs(Node* s1, Node* s2);
   void set_alignment(Node* s1, Node* s2, int align);
-  // Extend packset by following use->def and def->use links from pack members.
-  void extend_packset_with_more_pairs_by_following_use_and_def();
   int adjust_alignment_for_type_conversion(Node* s, Node* t, int align);
-  // Extend the packset by visiting operand definitions of nodes in pack p
-  bool follow_use_defs(Node_List* p);
-  // Extend the packset by visiting uses of nodes in pack p
-  bool follow_def_uses(Node_List* p);
-  // For extended packsets, ordinally arrange uses packset by major component
-  void order_def_uses(Node_List* p);
-  // Estimate the savings from executing s1 and s2 as a pack
-  int est_savings(Node* s1, Node* s2);
-  int adjacent_profit(Node* s1, Node* s2);
-  int pack_cost(int ct);
-  int unpack_cost(int ct);
-
-  // Combine packs A and B with A.last == B.first into A.first..,A.last,B.second,..B.last
-  void combine_pairs_to_longer_packs();
```
Note: moved and renamed some.

eme64 commented on src/hotspot/share/opto/superword.hpp:
```diff
@@ -57,16 +57,281 @@
+  GrowableArray<int> _right_to_left; // bb_idx -> bb_idx
+
+  // List of all left elements bb_idx, in the order of pair addition.
+  GrowableArray<int> _lefts_in_insertion_order;
```
Note: It turns out that the pair-extension is quite sensitive to the order in which pairs are found. Keeping the insertion order does something similar to a BFS. If this order is changed, then some of our examples don't pack properly. Well, they pack correctly on a local level, but then that prevents the proper packing further on.
Maybe I can remove the need for this order in the future, but for now I'll keep the order, the same order as the `_packset`'s pairs had.

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1214,197 +1183,206 @@ bool SuperWord::follow_def_uses(Node_List* p) {
+
+  // 1. Reduction
+  if (is_marked_reduction(use1) && is_marked_reduction(use2)) {
+    Node* first = use1->in(2);
```
Note: I also added some more cases for MulAddS2I test, reflecting some of the swaps in this method.

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1194,17 +1172,11 @@ bool SuperWord::follow_def_uses(Node_List* p) {
   }
-  if (num_s1_uses > 1) {
-    _race_possible = true;
-  }
```
Note: I removed `_race_possible` and `num_s1_uses`. We checked here if there was any node in a pack that has multiple uses. If that happens, it it spossible that `order_inputs_of_uses_to_match_def_pair` changes the order of inputs. This flag was set to fix everything again afterwards. But now I just call that algorithm always, implicitly always setting `_race_possible = true`. This does not cost so much, and makes things quite a bit simpler.

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1214,110 +1186,136 @@ bool SuperWord::follow_def_uses(Node_List* p) {
 
+  auto adjacent_profit = [&] (Node* s1, Node* s2) { return 2; };
+  auto pack_cost       = [&] (int ct) { return ct; };
+  auto unpack_cost     = [&] (int ct) { return ct; };
```
Note: I moved the methods from the `SuperWord` class to lambdas in this method, they were not used anywhere else.

eme64 commented on src/hotspot/share/opto/superword.hpp:
```diff
@@ -57,16 +57,281 @@
+private:
+  void set_pack(const Node* n, Node_List* pack) { _node_to_pack.at_put(_body.bb_idx(n), pack); }
+public:
+  Node_List* pack(const Node* n) const { return !_vloop.in_bb(n) ? nullptr : _node_to_pack.at(_body.bb_idx(n)); }
```
Note: replacement for `my_pack`.

eme64 commented on src/hotspot/share/opto/superword.hpp:
```diff
@@ -240,18 +505,13 @@ class SuperWord : public ResourceObj {
- public:
-  Node_List* my_pack(const Node* n)     const { return !in_bb(n) ? nullptr : _node_info.adr_at(bb_idx(n))->_my_pack; }
- private:
-  void set_my_pack(Node* n, Node_List* p)     { int i = bb_idx(n); grow_node_info(i); _node_info.adr_at(i)->_my_pack = p; }
```
Note: replaced with `PackSet.pack` and `PackSet.set_pack`.

eme64 commented on src/hotspot/share/opto/superword.hpp:
```diff
@@ -267,10 +527,8 @@ class SuperWord : public ResourceObj {
   // Can s1 and s2 be in a pack with s1 immediately preceding s2 and  s1 aligned at "align"
   bool stmts_can_pack(Node* s1, Node* s2, int align);
-  // Does s exist in a pack at position pos?
-  bool exists_at(Node* s, uint pos);
```
Note: replaced with `PairSet.has_left/right`.

eme64 commented on test/hotspot/jtreg/compiler/loopopts/superword/TestMulAddS2I.java:
```diff
@@ -133,4 +149,99 @@ public static int[] testc() {
+        for (int i = 0; i < ITER-2; i+=2) {
+            // Unrolled, with the same structure.
+            out[i+0] += ((sArr1[2*i+0] * sArr2[2*i+0]) + (sArr1[2*i+1] * sArr2[2*i+1]));
+            out[i+1] += ((sArr1[2*i+2] * sArr2[2*i+2]) + (sArr1[2*i+3] * sArr2[2*i+3]));
```
Note: extra cases to test the `MulAddS2I` input ordering in `order_inputs_of_uses_to_match_def_pair`.

chhagedorn commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1214,110 +1186,136 @@ bool SuperWord::follow_def_uses(Node_List* p) {
+//
+//    use1->in(i) == def1 || use2->in(def2)   ->    use1->in(i) == def1 && use2->in(def2)
+//
+// 3: Add/Mul (use1, use2): we can try to swap edges:
```
Is it not required for other nodes?

eme64 replied:
We can only do it for associative nodes like `Mul / Add`, and of course all the nodes that are subclasses of those (e.g. Max, Min, Or, etc). I can improve the comment, since that may not be immediately clear to the reader.

chhagedorn replied:
Thanks for the explanation. Yes, maybe you can clarify the comment :-)

chhagedorn commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1352,59 +1349,43 @@ int SuperWord::est_savings(Node* s1, Node* s2) {
+  // Iterate pair-chain by pair-chain, each from left-most to right-most.
+  Node_List* pack = nullptr;
+  for (PairSetIterator pair(_pairset); !pair.done(); pair.next()) {
+    Node* s1 = pair.left();
+    Node* s2 = pair.right();
```
Maybe you also want to name them `left` and `right` instead of `s1` and `s2`.

eme64 replied:
Just specifically here, or everywhere in the code?

chhagedorn replied:
Good question, we could probably also try to apply this renaming at other places. If it's not too much work, you can try it.

eme64 replied:
I did it in a few places, but not everywhere. Sometimes it is just easier to talk about `s1, s2` and `use1, use2` and `def1, def2` with numbers, rather than the much longer names using `left/right`.

chhagedorn commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1352,59 +1349,43 @@ int SuperWord::est_savings(Node* s1, Node* s2) {
+  for (PairSetIterator pair(_pairset); !pair.done(); pair.next()) {
+    Node* s1 = pair.left();
+    Node* s2 = pair.right();
+    if (_pairset.is_left_in_a_left_most_pair(s1)) {
```
Should we also assert here that `pack == nullptr` before creating the new list?

eme64 replied:
Good idea!

chhagedorn commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1352,59 +1349,43 @@ int SuperWord::est_savings(Node* s1, Node* s2) {
+      pack = new (arena()) Node_List(arena());
+      pack->push(s1);
+    }
+    pack->push(s2);
```
We should also assert that at this point, `pack != nullptr`.

eme64 replied:
Can do that, but I think it would just result in a nullptr exception anyway.
I always wonder if an assert makes the code more cluttered, so less readable, or if it states expectations more explicitly, which makes the code more readable :sweat_smile:

chhagedorn replied:
Right, well then you can leave it as is. One sometimes tends to be paranoid when it comes to using asserts in C2 :-)

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1049,30 +1034,37 @@ void SuperWord::set_alignment(Node* s1, Node* s2, int align) {
-// Extend packset by following use->def and def->use links from pack members.
-void SuperWord::extend_packset_with_more_pairs_by_following_use_and_def() {
+// Extend pairset by following use->def and def->use links from pair members.
+void SuperWord::extend_pairset_with_more_pairs_by_following_use_and_def() {
```
Not really. Inside we need lots of access to the SuperWord components (check `alignment`, `stmts_can_pack`, etc). You could always pass in a `SuperWord` reference, but that is not really nicer.
Maybe it would be easier in the future, once I make some other changes. But we will always need to have other information available (dependency graph, types, etc - these are all VLoopAnalyzer submodules), but for now we also rely on `alignment`, which I plan to remove, and lives in SuperWord.

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1214,110 +1186,136 @@ bool SuperWord::follow_def_uses(Node_List* p) {
+
+  // 1. Reduction
+  if (is_marked_reduction(use1) && is_marked_reduction(use2)) {
+    Node* first = use1->in(2);
```
I agree, it already was like that, and it disturbed it me too. I thought I'd leave it since I'll probably overhaul this whole code soon anyway. But I'll fix it already since it disturbs you too.

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1157,17 +1137,16 @@ bool SuperWord::follow_def_uses(Node_List* p) {
                   s1->_idx, align);
   }
 #endif
+  bool changed = false;
```
Good point!

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1214,110 +1186,136 @@ bool SuperWord::follow_def_uses(Node_List* p) {
+int SuperWord::estimate_cost_savings_when_packing_pair(const Node* s1, const Node* s2) const {
   int save_in = 2 - 1; // 2 operations per instruction in packed form
 
+  auto adjacent_profit = [&] (Node* s1, Node* s2) { return 2; };
```
Sure

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1214,110 +1186,136 @@ bool SuperWord::follow_def_uses(Node_List* p) {
+        // 4.a/b: MulAddS2I.
         if (i1 == 5 - i2) { // ((i1 == 3 && i2 == 2) || (i1 == 2 && i2 == 3) || (i1 == 1 && i2 == 4) || (i1 == 4 && i2 == 1))
-          u2->swap_edges(1, 2);
-          u2->swap_edges(3, 4);
+          use2->swap_edges(1, 2);
+          use2->swap_edges(3, 4);
         }
         if (i1 == 3 - i2 || i1 == 7 - i2) { // ((i1 == 1 && i2 == 2) || (i1 == 2 && i2 == 1) || (i1 == 3 && i2 == 4) || (i1 == 4 && i2 == 3))
```
I'd rather not spend too much time on this, since I'll overhaul it soon even more. I'm not touching the comments anyway ;)

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1214,110 +1186,136 @@ bool SuperWord::follow_def_uses(Node_List* p) {
+//          return PairOrderStatus::Unknown.
+//          Therefore, extend_pairset_with_more_pairs_by_following_use cannot extend to MulAddS2I,
+//          but there is a chance that extend_pairset_with_more_pairs_by_following_def can do it.
+//
```
Sounds good :)

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1214,110 +1186,136 @@ bool SuperWord::follow_def_uses(Node_List* p) {
+//
+// 2: Inputs of (use1, use2) already match (def1, def2), i.e. for all input indices i:
+//
+//    use1->in(i) == def1 || use2->in(def2)   ->    use1->in(i) == def1 && use2->in(def2)
```
Yes!

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1214,110 +1186,136 @@ bool SuperWord::follow_def_uses(Node_List* p) {
-//------------------------------est_savings---------------------------
 // Estimate the savings from executing s1 and s2 as a pack
-int SuperWord::est_savings(Node* s1, Node* s2) {
+int SuperWord::estimate_cost_savings_when_packing_pair(const Node* s1, const Node* s2) const {
```
done

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1329,18 +1327,17 @@ int SuperWord::est_savings(Node* s1, Node* s2) {
+
+    for (DUIterator_Fast kmax, k = s2->fast_outs(kmax); k < kmax; k++) {
+      if (use2 == s2->fast_out(k)) {
+        ct++;
```
Sure

eme64 commented:
@chhagedorn I think I have addressed all your remarks.
I agree I could have probably split the `PairSet` and `PackSet` into separate RFE's, that only became apparent once I did the work, and that there is not really any shared code between the two. Though I'm not sure it is worth splitting this up now, after all they are really about the `_packset`.

chhagedorn commented on src/hotspot/share/opto/superword.hpp:
```diff
@@ -57,16 +57,281 @@
+  const VLoop& _vloop;
+  const VLoopBody& _body;
+
+  // The "packset" proper: an array of "packs"
```
What do you mean by " The "packset" proper"?

eme64 replied:
Changed comment to `// Set of all packs:`

chhagedorn commented on src/hotspot/share/opto/superword.hpp:
```diff
@@ -219,9 +485,8 @@ class SuperWord : public ResourceObj {
 
   bool     do_vector_loop()        { return _do_vector_loop; }
 
-  const GrowableArray<Node_List*>& packset() const { return _packset; }
+  const PackSet& packset()   const { return _packset; }
```
Somehow a strange alignment. You might want to fix that.

eme64 replied:
Had it aligned with `do_vector_loop`. But it seems unnecessary, changed it.

chhagedorn commented on test/hotspot/jtreg/compiler/loopopts/superword/TestMulAddS2I.java:
```diff
@@ -36,14 +36,20 @@
 import jdk.test.lib.Platform;
 
 public class TestMulAddS2I {
```
Might be worth to add 8325252 as `@bug` number as well since you added quite a few tests.

chhagedorn reviewed (commented):
Thanks for the updates. The PR generally looks good! I now fully reviewed it and left some more comments.

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1352,59 +1354,45 @@ int SuperWord::est_savings(Node* s1, Node* s2) {
-                                             Node_List* pack,
-                                             SplitTask task)
+SplitStatus PackSet::split_pack(const char* split_name,
+                                Node_List* pack,
```
I will keep that in mind. It could really improve readability. Let's do it in a separate RFE!

chhagedorn replied:
Sounds good!

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -1511,130 +1499,110 @@ SuperWord::SplitStatus SuperWord::split_pack(const char* split_name,
 }
 
 // Split packs at boundaries where left and right have different use or def packs.
 void SuperWord::split_packs_at_use_def_boundaries() {
```
Yes, let's consider that for a future RFE.

But for now I can say this:
We moved the implementation of `split / filter` to the packset, and that makes sense: we can hide the packset internals, and don't have them spill out to the SuperWord code. We can just call `_packset.split_packs` with a `split_strategy`. But defining the split strategy itself depends on other SuperWord components, and so I think conceptually they belong with SuperWord. They query `reductions`, `dependency graph`, `implemented`, `AlignmentSolution`, `profitable`, etc.
Sure, we can just pass a SuperWord reference, but that does not really seem right to me either. For me, the SuperWord class is there to manage the interface between all the components, and try to avoid passing components to other components, wherever possible. So I would rather have a list of methods in SuperWord, and each such method defines how the components interact (e.g. packset and alignment, packset and AlignmentSolution, pairset and packset, ...).
But we can discuss this further, and maybe come up with an even better solution. My hope is just that we separate the components as much as possible, so that we know that only a handful of them interact at a given point. That makes the whole beast more managable.

chhagedorn replied:
That's a valid point. Let's have another look at `SuperWord`, `PairSet` and `PackSet` once the code stabilizes to check whether it needs improvements or not by moving some code to the `*Set` classes.

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -2141,7 +2081,7 @@ class PacksetGraph {
   // Create nodes (from packs and scalar-nodes), and add edges, based on the dependency graph.
   void build() {
-    const GrowableArray<Node_List*>& packset = _slp->packset();
+    const PackSet& packset = _slp->packset();
```
You are right, I could actually remove the SuperWord reference from the `PacksetGraph` completely, and just pass the components, as `const` references. That would be really nice.

If it is ok for you, I will do this in a future RFE.

Actually, I plan to completely overhaul the `PacksetGraph`. It will be transformed to the `VTransformGraph`, and it will do:
- Cycle checking (like today)
- Evaluate the cost-model
- Execute: each node knows how to replace its packed scalar nodes with vector nodes (basically refactoring `SuperWord::output` away)
- etc.

We may even be able to take the `VTransformGraph` and try to widen all nodes, or make transformations on this graph, a simplified version of IGVN. I have lots of ideas that would be unlocked with this new graph-based approach.

chhagedorn replied:
Sounds exciting! 

> If it is ok for you, I will do this in a future RFE.

Sure, let's leave this as it is now.

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -2461,7 +2401,7 @@ bool SuperWord::output() {
   for (int i = 0; i < body().length(); i++) {
     Node* n = body().at(i);
-    Node_List* p = my_pack(n);
+    Node_List* p = _packset.pack(n);
```
Hmm. I tried it, but then we often also have a variable `pack`. So I now changed the `pack(n)` to `get_pack(n)`. I think that is better anyway, it suggests that we "get" something, rather than "pack" something.

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -3822,28 +3719,45 @@ void SuperWord::adjust_pre_loop_limit_to_align_main_loop_vectors() {
-void SuperWord::print_pack(Node_List* p) {
-  for (uint i = 0; i < p->size(); i++) {
-    print_stmt(p->at(i));
+void PackSet::print_pack(Node_List* pack) const {
```
This is where a `Pack` class would eventually come in quite nicely. But I will change it to static for now.

eme64 commented on src/hotspot/share/opto/superword.hpp:
```diff
@@ -57,16 +57,281 @@
+  const VLoop& _vloop;
+  const VLoopBody& _body;
+
+  // Doubly-linked pairs. If not linked: -1
+  GrowableArray<int> _left_to_right; // bb_idx -> bb_idx
+  GrowableArray<int> _right_to_left; // bb_idx -> bb_idx
```
Great idea, I made a slightly more complex example, inspired by yours!

chhagedorn replied:
Cool, thanks for adding an even more extensive example. I like it and helps with understanding the idea better :-)

eme64 commented:
@chhagedorn thanks very much for the review, there were some great suggestions! I fixed almost all, and left some responses at the others.

I'll repeat them here:

------------

You are right, I could actually remove the SuperWord reference from the PacksetGraph completely, and just pass the components, as const references. That would be really nice.

If it is ok for you, I will do this in a future RFE.

Actually, I plan to completely overhaul the PacksetGraph. It will be transformed to the VTransformGraph, and it will do:

Cycle checking (like today)
Evaluate the cost-model
Execute: each node knows how to replace its packed scalar nodes with vector nodes (basically refactoring SuperWord::output away)
etc.
We may even be able to take the VTransformGraph and try to widen all nodes, or make transformations on this graph, a simplified version of IGVN. I have lots of ideas that would be unlocked with this new graph-based approach.

------------

We moved the implementation of split / filter to the packset, and that makes sense: we can hide the packset internals, and don't have them spill out to the SuperWord code. We can just call _packset.split_packs with a split_strategy. But defining the split strategy itself depends on other SuperWord components, and so I think conceptually they belong with SuperWord. They query reductions, dependency graph, implemented, AlignmentSolution, profitable, etc.
Sure, we can just pass a SuperWord reference, but that does not really seem right to me either. For me, the SuperWord class is there to manage the interface between all the components, and try to avoid passing components to other components, wherever possible. So I would rather have a list of methods in SuperWord, and each such method defines how the components interact (e.g. packset and alignment, packset and AlignmentSolution, pairset and packset, ...).
But we can discuss this further, and maybe come up with an even better solution. My hope is just that we separate the components as much as possible, so that we know that only a handful of them interact at a given point. That makes the whole beast more managable.

----------

Using `Pack` instead of `Node_List`. Great idea, would improve readability. Let's do it in a future RFE.

chhagedorn commented on src/hotspot/share/opto/superword.hpp:
```diff
@@ -288,16 +331,34 @@ class PackSet : public StackObj {
 
 private:
-  void set_pack(const Node* n, Node_List* pack) { _node_to_pack.at_put(_body.bb_idx(n), pack); }
+  void map_node_in_pack(const Node* n, Node_List* new_pack) {
+    assert(get_pack(n) == nullptr, "was previously unmapped");
+    _node_to_pack.at_put(_body.bb_idx(n), new_pack);
+  }
+
+  void remap_node_in_pack(const Node* n, Node_List* new_pack) {
+    assert(get_pack(n) != nullptr && new_pack != nullptr && get_pack(n) != new_pack, "was previously mapped");
+    _node_to_pack.at_put(_body.bb_idx(n), new_pack);
+  }
+
+  void unmap_node_in_pack(const Node* n) {
+    assert(get_pack(n) != nullptr, "was previously mapped");
+    _node_to_pack.at_put(_body.bb_idx(n), nullptr);
+  }
+
+  void unmap_all_nodes_in_pack(Node_List* old_pack) {
+    for (uint i = 0; i < old_pack->size(); i++) {
+      unmap_node_in_pack(old_pack->at(i));
+    }
+  }
```
You even added a `remap` method and added some asserts, nice!

chhagedorn reviewed (approved):
Thanks for doing all the updates. It looks good to me now!

vnkozlov reviewed (approved):
Good.

eme64 commented:
Thanks @chhagedorn for the detailed review, thanks also @vnkozlov !
/integrate
